all: Replace MP_STATE_PORT with MP_STATE_ROOT.

jimmo opened the pull request:
Implements option (3) from #12228. cc @robert-hh @dlech 

---

Before MP_REGISTER_ROOT_POINTER was added, a port could define additional
root pointers in its mpconfigport.h. It made sense to refer to them via a
distinct macro, hence MP_STATE_PORT, even though all ports alias that to
MP_STATE_VM.

Now the MP_REGISTER_ROOT_POINTER feature can be used to implement root
pointers for modules that mpstate.h does not know about, such as modules in
extmod, or user c module. We could also consider moving some
config-optional stuff that's currently in mpstate.h to use
MP_REGISTER_ROOT_POINTER. But things defined with MP_REGISTER_ROOT_POINTER
are no longer "port". So for example, in modbluetooth_nimble.c we use
MP_STATE_PORT(bluetooth_nimble_root_pointers), but the only thing that's
"port" about them is that bluetooth can be enabled by the port/board.

This renames all previous uses of MP_STATE_PORT to the new MP_STATE_ROOT.

Furthermore, it moves the generated entries out of the mp_state_vm_t
structure and into an new mp_state_root_t structure that sits between
thread and vm (so it's still part of root pointer searching).

Also fixes several cases where MP_STATE_VM was incorrectly used for what
should be now MP_STATE_ROOT.

Other than the logical improvement, it also means that ports no longer
have to define MP_STATE_PORT (one less thing to duplicate in
mpconfigport.h).

_This work was funded through GitHub Sponsors._

jimmo committed:
all: Replace MP_STATE_PORT with MP_ROOT_POINTER.

Before MP_REGISTER_ROOT_POINTER was added, a port could define additional
root pointers in its mpconfigport.h. It made sense to refer to them via a
distinct macro, hence MP_STATE_PORT, even though all ports alias that to
MP_STATE_VM.

Now the MP_REGISTER_ROOT_POINTER feature can be used to implement root
pointers for modules that mpstate.h does not know about, such as modules in
extmod, or user c module. We could also consider moving some
config-optional stuff that's currently in mpstate.h to use
MP_REGISTER_ROOT_POINTER. But things defined with MP_REGISTER_ROOT_POINTER
are no longer "port". So for example, in modbluetooth_nimble.c we use
MP_STATE_PORT(bluetooth_nimble_root_pointers), but the only thing that's
"port" about them is that bluetooth can be enabled by the port/board.

This renames all previous uses of MP_STATE_PORT to the new MP_ROOT_POINTER.
The name is chosen to match MP_REGISTER_ROOT_POINTER.

Furthermore, it moves the generated entries out of the mp_state_vm_t
structure and into an new mp_state_root_t structure that is nested in
mp_state_vm_t, but all the offsets are the same so no code size diff.

Also fixes several cases where MP_STATE_VM was incorrectly used for what
should be now MP_ROOT_POINTER.

Other than the logical improvement, it also means that ports no longer
have to define MP_STATE_PORT (one less thing to duplicate in
mpconfigport.h).

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dlech commented:
I think `MP_ROOT_POINTER` would be a better name than `MP_STATE_ROOT` since it pairs nicer with `MP_REGISTER_ROOT_POINTER`. And the name change could still be nice even if we don't end up creating a new struct and  `MP_ROOT_POINTER` is just an alias for `MP_STATE_VM`.

jimmo commented:
Thanks @dlech, sounds reasonable to me. I have updated the PR. Could also consider `MP_STATE_ROOT_POINTER`...

Also, I realised that I could get the best of both worlds by nesting the `mp_state_root_t` struct inside `mp_state_vm_t` at the original location. This gets us back to the original code size.

dlech reviewed (approved):
LGTM

(If you want to do some more housekeeping we could make the use of `[]` inside or outside the macro `()` consistent. It looks like we have it about half and half currently.)

jimmo committed:
all: Standardise on MP_ROOT_POINTER(x)[i].

For registered root pointers declared as
`MP_REGISTER_ROOT_POINTER(mp_obj_t x[N]);` it's valid to use both
`MP_ROOT_POINTER(x)[i]` and `MP_ROOT_POINTER(x[i])`.

For consistency, standardise on the former everywhere.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
> (If you want to do some more housekeeping we could make the use of `[]` inside or outside the macro `()` consistent. It looks like we have it about half and half currently.)

I do love me some housekeeping.

I've added a commit that standardises on `MP_ROOT_POINTER(x)[i]` everywhere. Chose that option because
a) It seems to be slightly more common (but it's close, 111 vs 72 occurrences)
b) It seems slightly more correct to me. e.g. because there are places that do `memset(MP_ROOT_POINTER(x), ...)`.
